server/src/main/java/org/elasticsearch/search/fetch/subphase/ExplainPhase.java

```diff
@@ -32,24 +33,27 @@
 public final class ExplainPhase implements FetchSubPhase {
 
     @Override
-    public void hitExecute(SearchContext context, HitContext hitContext) {
+    public FetchSubPhaseExecutor getExecutor(SearchContext context) {
         if (context.explain() == false || context.hasOnlySuggest()) {
-            return;
+            return null;
         }
-        try {
-            final int topLevelDocId = hitContext.hit().docId();
-            Explanation explanation = context.searcher().explain(context.query(), topLevelDocId);
+        return new FetchSubPhaseExecutor() {
+            @Override
+            public void setNextReader(LeafReaderContext readerContext) {
 
-            for (RescoreContext rescore : context.rescore()) {
-                explanation = rescore.rescorer().explain(topLevelDocId, context.searcher(), rescore, explanation);
             }
-            // we use the top level doc id, since we work with the top level searcher
-            hitContext.hit().explanation(explanation);
-        } catch (IOException e) {
```

Comment:
The `finally` block is a no-op (nowhere in the code do we add a releasable for `Lifetime.COLLECTION` and it will be removed entirely by the `ReaderContext` refactor) so it's safe to remove.  Given that `execute` is declared as `throws IOException` it seems weird to catch `IOException` and rethrow as something else; but perhaps `FetchPhase` itself should throw a `FetchPhaseExecutionException`?